Add presigned URLs to erlcloud

lbakerchef opened the pull request:
This PR is part of a broader effort to replace AWS sigv2 with sigv4, encompassing modifications to oc_erchef, bookshelf, mini_s3, erlcloud, pedant clients, and other systems. Erlcloud received significant modifications for presigning of URLs, date and time mocking/modification, injection of host headers and ports, presigning with a' la carte headers, and expiration windows, among others. New unit tests exercising various aspects of newly-added functionality were also added.

Should this be merged, or (more likely) **left as its own branch**?  Unless someone believes otherwise, I'm recommending leaving this as a separate branch.  

See:
https://github.com/chef/chef-server/issues/1911
https://github.com/chef/chef-server/pull/1993
chef/mini_s3#31

- [x] Green verify pipeline
- [x] Green omnibus/adhoc pipeline
- [x] Green integration test pipeline
- [x] Green dialyzer pass

tommysrn committed:
Add functions to create pre-signed URLs for S3

erlcloud_s3:make_presigned_v4_url.
See https://github.com/erlcloud/erlcloud/issues/562

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
Add eunit tests for presigned URLs

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
replace rebar.lock

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
Add capability to generate a signature using a passed-in date

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
Use date header in sign_v4 if available

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
More consistent treatment of host header

1) Host headers are treated inconsistently between various componenents
(erlcloud, mini_s3, bookshelf, etc).

2) The generated host header was incorrect in certain cases, eg the host
header generated from http://127.0.0.1:4321 left off the port and resulted
in Host: 127.0.0.1.  However host header specs say that in the case of a
missing port the assumed port will be based on the scheme - 80 for http, 443
for https.  So the port should be added in this case, as it is 4321.

https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
Add iso_8601_basic_time/1 function

This functionality was added to support creation of expiration windows
needed in oc_erchef and other components.

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
Fix fips issue on rhel boxes

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
Add capability to presign URLs with passed-in host header and date

While processing a sigv4 presigned url request, Bookshelf needs to
be able to determine whether the presigned url is valid, appropriately
signed, etc.

The only way I know to do this is to create a new presigned url
using whatever inputs the original presigned url was created with, and
compare the two presigned urls.  In order to do this, I have to use the
same host header and date the original presigned url was created with.
But it wasn't possible to pass in a host header and date, thus this commit.

Signed-off-by: Lincoln Baker <lbaker@chef.io>

code cleanup src/erlcloud_s3.erl

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
Add capability to generate presigned urls with a' la carte headers

Signed-off-by: Lincoln Baker <lbaker@chef.io>

stevendanna commented on src/erlcloud_aws.erl:
```diff
@@ -989,22 +996,41 @@ sign_v4_headers(Config, Headers, Payload, Region, Service) ->
+    Headers1 =
+        case proplists:get_value("x-amz-date", Headers0) of
+            undefined ->
+                Date = Date0,
```
Is this unused?

lbakerchef replied:
Date? Headers1? They are used.

stevendanna commented on src/erlcloud_aws.erl:
```diff
@@ -989,22 +996,41 @@ sign_v4_headers(Config, Headers, Payload, Region, Service) ->
+
+-spec sign_v4(atom(), list(), aws_config(), headers(), string() | binary(), string(), string(), list(), string()) -> headers().
+sign_v4(Method, Uri, Config, Headers0, Payload, Region, Service, QueryParams, Date0) ->
+    % use passed-in x-amz-date header or create one
```
I find the semantics of this a _little_ strange in that we are passing an explicit date but the x-amz-date command overrides this. I assume this is because of the double-duty this is doing for verifying our signatures (which will likely have the date header coming in via Headers) and generating new signatures (which I am assuming won't have the date header coming in via Headers)

lbakerchef replied:
I don't know if this helps, but the way it works is: you pass in a date, but if there is a date header also being passed in, the date in the date header is used.  If not, the passed-in date is used.

lbakerchef replied:
I think your 'double-duty' comment is probably correct.

stevendanna commented on src/erlcloud_s3.erl:
```diff
@@ -1084,17 +1088,95 @@ make_link(Expire_time, BucketName, Key, Config) ->
+  ToSign = erlcloud_aws:to_sign(Date, CredentialScope, CanonicalRequest),
+  SigningKey = erlcloud_aws:signing_key(Config, Date, Region, Service),
+  [Result] = erlcloud_aws:base16(erlcloud_util:sha256_mac(SigningKey, ToSign)),
+  Result.
```
I don't see many tests for this function. It might be nice to generate some pre-signed URLs with another tool and then use the results of that as test data that we could use to validate this function.

lbakerchef replied:
I don't think this is my code.  Would you still like me to add tests for it?

PrajaktaPurohit replied:
yeah - that would be really nice if we can add tests, that might also make it easier to get the commit merged upstream.

lbakerchef replied:
"...generate some pre-signed URLs with another tool..."

1) Does anyone know of a trusted tool that does this?

2) In theory we could simply take presigned URLs that worked against S3 in the past (meaning, we know they are valid) and run them through again, but the problem is, they expire.  The date looks mockable so that might not be an issue.  I'd have to create a build that spits out the inputs+outputs that came into/out-of this function which were then subsequently used to hit S3 with presigned URLs successfully.

lbakerchef replied:
Ok, I did #2 above.  I created a build to log diagnostics, and ran `chef-server-ctl test` against S3.  I was able to get the inputs and outputs to the signature function to create tests.  The tests worked.  However, my S3 `access_key_id` is in the source code.  Perhaps a better approach is to run against Bookshelf instead, although that is using our own code to verify our own code.

lbakerchef replied:
Rewrote to use Bookshelf vs S3.

stevendanna reviewed (commented):
Overall this one looks good to me.  I'd like to see a few more tests and I think we might need to clean up a bit how we handle the Host headers; but I'd like to see the call sites before making that call.

PrajaktaPurohit commented on rebar.config:
```diff
@@ -2,7 +2,7 @@
 {erl_opts,
  [debug_info,
-  {platform_define, "^19.|^2", 'ERLANG_OTP_VERSION_19'},
+%  {platform_define, "^19.|^2", 'ERLANG_OTP_VERSION_19'},
```
We are currently on 22.

lbakerchef replied:
Can we just remove the comment, or do you want something like:
`{platform_define, "^22", 'ERLANG_OTP_VERSION_22'},`

lbakerchef replied:
Also, if we plan to try to merge into their library in the future, do we want to change this?  They may want to support more platforms?

lbakerchef replied:
Done.

lbakerchef replied:
I think the line restricts platforms to OTP 19.  I removed it so that it will run on whatever platform.  I tested it on 22 and 24.

lbakerchef committed:
add eunit tests for signature/8

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
Upgrade to Erlang 24x. (#4)

Fix license_scout error in Bookshelf.

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
fix undefined type error.

===> Compiling erlcloud
===> Compiling src/erlcloud_aws.erl failed
src/erlcloud_aws.erl:1152:118: type headers() undefined

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
fix random:uniform deprecation.

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
fix unused variable warning

===> Compiling src/erlcloud_s3.erl failed
src/erlcloud_s3.erl:1141:22: variable 'SignedHeaders' is unused

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
fix warning - variable exported from case

===> Compiling src/erlcloud_aws.erl failed
src/erlcloud_aws.erl:1170:40: variable 'Date' exported from 'case' (line 1155, column 9)

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
fix tests

Signed-off-by: Lincoln Baker <lbaker@chef.io>

lbakerchef committed:
remove platform restrictions

Signed-off-by: Lincoln Baker <lbaker@chef.io>

vinay-satish reviewed (approved):
Code looks good to me, could you please add the links to the build kite pipelines.

lbakerchef commented:
> Code looks good to me, could you please add the links to the build kite pipelines.

@vinay-satish See this PR for adhocs and umbrellas:
https://github.com/chef/chef-server/pull/2097

marcparadise commented on src/erlcloud_s3.erl:
```diff
@@ -506,7 +511,7 @@ put_bucket_lifecycle(BucketName, Policy, Config)
 put_bucket_lifecycle(BucketName, XmlPolicy, Config)
   when is_list(BucketName), is_binary(XmlPolicy), is_record(Config, aws_config) ->
-    Md5 = base64:encode(crypto:hash(md5, XmlPolicy)),
+    Md5 = base64:encode(erlang:md5(XmlPolicy)),
```
Why the change from `crypto:hash` to `erlang:md5`? I distantly remember we had to go in the other direction once, years ago.

marcparadise commented on src/erlcloud_s3.erl:
```diff
@@ -506,7 +511,7 @@ put_bucket_lifecycle(BucketName, Policy, Config)
 put_bucket_lifecycle(BucketName, XmlPolicy, Config)
   when is_list(BucketName), is_binary(XmlPolicy), is_record(Config, aws_config) ->
-    Md5 = base64:encode(crypto:hash(md5, XmlPolicy)),
+    Md5 = base64:encode(erlang:md5(XmlPolicy)),
```
Why the change from crypto:hash to erlang:md5? I distantly remember we had to go in the other direction once, years ago.

lbakerchef replied:
The commit says that it is to fix a FIPS issue on rhel boxes.  If i recall correctly, @PrajaktaPurohit may know the context about this and know which one is correct - the `crypto:hash` or `erlang:md5`.
